File: JS/day_7/part1.js
import { readFile } from "node:fs/promises";

async function part1(path) {
  const file = await readFile(path, "utf8");
  const data = file.trim().split("\n");

  let count = 0;

  let beams = Array(data[0].length).fill(false);

  const start = data[0].indexOf("S");
  beams[start] = true;

  let next = Array(beams.length).fill(false);

  for (const line of data.slice(1)) {
    next.fill(false);

    beams.forEach((active, col) => {
      if (!active) return;

      if (line[col] === "^") {
        if (col - 1 >= 0) next[col - 1] = true;
        if (col + 1 < line.length) next[col + 1] = true;

        count++;
      } else {
        next[col] = true;
      }
    });

    [beams, next] = [next, beams];
  }

  console.log(count);
}

part1("input.txt");

File: GO/day_7/part1.go
package day7

import (
	"bytes"
	"fmt"

	"aoc2025/utils"
)

func Part1(f string) {
	data := utils.ReadByteLines(path[f])
	count := 0

	beams := make([]bool, len(data[0]))
	beams[bytes.Index(data[0], []byte("S"))] = true

	next := make([]bool, len(beams))

	for _, line := range data[1:] {
		for i := range next {
			next[i] = false
		}

		for col, ok := range beams {
			if !ok {
				continue
			}

			switch line[col] {
			case '^':
				count++
				next[col+1] = true
				next[col-1] = true
			default:
				next[col] = true
			}
		}

		beams, next = next, beams
	}

	fmt.Printf("Day 7 | Part 1: %d\n", count)
}
